docs/source/dev/design.rst

```diff
@@ -0,0 +1,99 @@
+
+Backend
+-------
+One of the main goals of this library is to provide a high-level common interface to synchronize persistent data with a
+volatile representation in memory. This includes handling data in any number of supported file formats transparently.
+To enable the user to handle hierarchical, self-describing file formats while disregarding the actual nitty-gritty
+details of just those file formats, the possible operations have been reduced to a common set of
+`IOTasks <https://github.com/openPMD/openPMD-api/blob/dev/include/openPMD/IO/IOTask.hpp>`:
+
+.. literalinclude:: IOTask.hpp
+   :lines: 41-66
+
+Every task is designed to be a fully self-contained description of one such atomic operation. By describing a required
+minimal step of work (without any side-effect), these operations are the foundation of the unified handling mechanism
+across suitable file formats.
+The actual low-level exchange of data is implemented in IOHandlers, one per file format (possibly two if handling
+MPI-parallel work is possible and requires different behaviour).
+The only task of these IOHandlers is to execute one atomic IOTask at a time. Ideally, additional logic is contained to
+improve performance by keeping track of open file handles, deferring and coalescing parts of work, avoiding redundant
+operations. It should be noted that while this is desirable, sequential consistency must be guaranteed (see
+:ref:`_queue_label`.)
+
+`AbstractParameter` and subclasses as typesafe descriptions of task parameters
+
+`WRITABLE` as unique identification in task, corresponding to node in frontend hierarchy (tree-like structure)
+
+subclass of `AbstractIOHandler` to ensure simple extensibilty
+
+only two public interface methods (`enqueue()` and `flush()`) to hide separate behaviour & state
+
+`AbstractFilePosition` as a format-dependent location inside persistent data (e.g. node-id / path string)
+
+should be entirely agnostic to openPMD and just treat transferred data as raw bytes without *knowledge*
+
+.. _queue_label:
+I/O-Queue
+---------
+To keep coupling between openPMD logic and actual low-level I/O to a minimum, a sequence of atomic I/O-Tasks is used to
+transfer data between logical and physical representation. Individual tasks are scheduled by frontend application logic
+and stored in a data structure that allows for FIFO order processing (in future releases, this order might be relaxed).
+Tasks are not executed during their creation, but are instead buffered in this queue. Disk accesses can be coalesced and
+high access latencies can be amortized by performing multiple tasks bunched together.
+At appropriate points in time, the used backend processes all pending tasks (strict, single-threaded, synchronous FIFO
+is currently used in all backends, but is not mandatory as long as consistency with that order can be guaranteed).
+
+A typical sequence of tasks that are scheduled during the read of an existing file *could* look something like this:
+
+``
```

Comment:
You can just use a simple [source block](http://www.sphinx-doc.org/en/stable/rest.html#source-code) here